src/loops.rs

```diff
@@ -242,3 +276,172 @@ fn is_break_expr(expr: &Expr) -> bool {
+    depth: u32,              // depth of conditional expressions
+    done: bool
+}
+
+impl<'v, 't> Visitor<'v> for InitializeVisitor<'v, 't> {
+    fn visit_decl(&mut self, decl: &'v Decl) {
+        // Look for declarations of the variable
+        if let DeclLocal(ref local) = decl.node {
+            if local.pat.id == self.var_id {
+                if let PatIdent(_, ref ident, _) = local.pat.node {
+                    self.name = Some(ident.node.name);
+
+                    self.state = if let Some(ref init) = local.init {
+                        if is_integer_literal(init, 0) {
+                            VarState::Warn
+                        } else {
+                            VarState::Declared
+                        }
+                    }
+                    else {
+                        VarState::Declared
+                    }
+                }
+            }
+        }
+        walk_decl(self, decl);
+    }
+
+    fn visit_expr(&mut self, expr: &'v Expr) {
+        if self.state == VarState::DontWarn || expr == self.end_expr {
+            self.done = true;
+        }
+        // No need to visit expressions before the variable is
+        // declared or after we've rejected it.
+        if self.state == VarState::IncrOnce || self.done {
+            return;
+        }
+
+        // If node is the desired variable, see how it's used
+        if var_def_id(self.cx, expr) == Some(self.var_id) {
+            if let Some(parent) = get_parent_expr(self.cx, expr) {
+                match parent.node {
+                    ExprAssignOp(_, ref lhs, _) if lhs.id == expr.id => {
+                        self.state = VarState::DontWarn;
+                    },
+                    ExprAssign(ref lhs, ref rhs) if lhs.id == expr.id => {
+                        self.state = if is_integer_literal(rhs, 0) && self.depth == 0 {
+                            VarState::Warn
+                        } else {
+                            VarState::DontWarn
+                        }},
+                    _ => ()
+                }
+            }
+        }
+        // If there are other loops between the declaration and the target loop, give up
+        else if is_loop(expr) {
+            self.state = VarState::DontWarn;
+            self.done = true;
+            return;
+        }
+        // Keep track of whether we're inside a conditional expression
+        else if is_conditional(expr) {
+            self.depth += 1;
+            walk_expr(self, expr);
+            self.depth -= 1;
+            return;
+        }
+        walk_expr(self, expr);
+    }
+}
+
+fn var_def_id(cx: &Context, expr: &Expr) -> Option<NodeId> {
+    if let Some(path_res) = cx.tcx.def_map.borrow().get(&expr.id) {
+        if let DefLocal(node_id) = path_res.base_def {
+            return Some(node_id)
+        }
+    }
+    None
+}
+
+fn is_loop(expr: &Expr) -> bool {
+    match expr.node {
+        ExprLoop(_,_) | ExprWhile(_,_,_)  => true,
```

Comment:
Thanks, I was hoping there was a syntax for that but I couldn't figure out what it was.